xapian-applications/omega/handler_gmime.cc

```diff
@@ -0,0 +1,174 @@
+/** @file handler_gmime.cc
+ * @brief Extract text and metadata using gmime.
+ */
+/* Copyright (C) 2019 Bruno Baruffaldi
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include "myhtmlparse.h"
+#include <string.h>
+
+#include <glib.h>
+#include <gmime/gmime.h>
+
+using namespace std;
+
+const unsigned SIZE = 8192;
+
+static void
+extract_html(const string& text, string& charset, string& dump)
+{
+    MyHtmlParser parser;
+    if (charset.empty())
+	charset = "UTF-8";
+    try {
+	parser.ignore_metarobots();
+	parser.parse_html(text, charset, false);
+	dump += parser.dump;
+    } catch (const string& newcharset) {
+	parser.reset();
+	parser.ignore_metarobots();
+	parser.parse_html(text, newcharset, true);
+	dump += parser.dump;
+    } catch (...) {
+	return;
+    }
+}
+
+size_t decode(unsigned char* text, size_t len, GMimeContentEncoding encoding)
+{
+    unsigned char buffer[SIZE];
+    int state;
+    guint32 save;
```

Comment:
I think you probably need to be preserving the values of `state` and `save` between calls to this function - otherwise it's not going to handle encoded data which is split between two chunks.

Consider quoted-printable sequence `=46` where `=4` is at the end of one chunk and `6` at the start of the next.  Something needs to remember that we're in the middle of a QP sequence and that we already saw `4` as the first digit.  I'd imagine currently this will wrongly decode as `6` instead of `F`.

Reply:
Actually, I am not sure about how it works. I could not find any example and the manual doesn't say much about it.